File: BackupUtility/Classes/FileIgnore/IgnoreOnExtensionsChecker.vb
Imports System.IO
Imports GenericClassLibrary.FileSystem

Public Class IgnoreOnExtensionsChecker
    Implements IFileIgnoreChecker

    Private ReadOnly _fileSystem As IFileSystem
    Private ReadOnly _exclusiveExt As String()
    Private ReadOnly _exceptionHandler As IExceptionHandler

    Public Sub New(exclusiveExt As String(), fileSystem As IFileSystem, exceptionHandler As IExceptionHandler)
        _fileSystem = fileSystem
        _exceptionHandler = exceptionHandler
        _exclusiveExt = exclusiveExt
    End Sub

    Public Function IgnoreFile(file As String) As Boolean Implements IFileIgnoreChecker.IgnoreFile
        Try
            If _fileSystem.File.GetExtension(file).Length > 0 Then
                If _exclusiveExt.Contains(_fileSystem.File.GetExtension(file).Substring(1)) Then Return True
            End If
        Catch ex As UnauthorizedAccessException
            _exceptionHandler.HandleException("Copy File: " & file & " failed(" & ex.Message & ")", ex)
        Catch ex As IOException
            _exceptionHandler.HandleException("Copy File: " & file & " failed(" & ex.Message & ")", ex)
        End Try
        Return False
    End Function
End Class

File: BackupUtility/Classes/FileIgnore/IgnoreOnExtensionsChecker.cs
using GenericClassLibrary.FileSystem;
using System;
using System.IO;
using System.Linq;
using ZCopy.Interfaces;

namespace ZCopy.Classes.FileIgnore
{
    public class IgnoreOnExtensionsChecker : IFileIgnoreChecker
    {
        private readonly IFileSystem _fileSystem;
        private readonly string[] _exclusiveExt;
        private readonly IExceptionHandler _exceptionHandler;

        public IgnoreOnExtensionsChecker(string[] exclusiveExt, IFileSystem fileSystem, IExceptionHandler exceptionHandler)
        {
            _fileSystem = fileSystem;
            _exceptionHandler = exceptionHandler;
            if (exclusiveExt != null)
            {
                _exclusiveExt = exclusiveExt.Select(a => a.ToLower()).ToArray();
            }
            else
            {
                _exclusiveExt = new string[0];
            }
        }

        public bool IgnoreFile(FolderMap baseMap, string file)
        {
            try
            {
                if (_exclusiveExt.Length == 0)
                {
                    return false;
                }
                string fileExtension = _fileSystem.File.GetExtension(file).ToLower();
                if (fileExtension.Length > 0 && _exclusiveExt.Contains(fileExtension.Substring(1)))
                {
                    return true;
                }
            }
            catch (UnauthorizedAccessException ex)
            {
                _exceptionHandler.HandleException("Copy File: " + file + " failed(" + ex.Message + ")", ex);
            }
            catch (IOException ex)
            {
                _exceptionHandler.HandleException("Copy File: " + file + " failed(" + ex.Message + ")", ex);
            }
            return false;
        }
    }
}
